server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java

```diff
@@ -68,147 +66,171 @@
         this.mimeMessageConverter = mimeMessageConverter;
         this.mailSpool = mailSpool;
         this.mailFactory = mailFactory;
+        this.systemMailboxesProvider = systemMailboxesProvider;
     }
 
     @Override
     public SetMessagesResponse process(SetMessagesRequest request, MailboxSession mailboxSession) {
-        Mailbox<Id> outbox;
-        try {
-            outbox = getOutbox(mailboxSession).orElseThrow(() -> new MailboxRoleNotFoundException(Role.OUTBOX));
-        } catch (MailboxException | MailboxRoleNotFoundException e) {
-            LOGGER.error("Unable to find a mailbox with role 'outbox'!");
-            throw Throwables.propagate(e);
-        }
-
-        // handle errors
-        Predicate<CreationMessage> validMessagesTester = CreationMessage::isValid;
-        Predicate<CreationMessage> invalidMessagesTester = validMessagesTester.negate();
-        SetMessagesResponse.Builder responseBuilder = SetMessagesResponse.builder()
-                .notCreated(handleCreationErrors(invalidMessagesTester, request));
-
-        return request.getCreate().entrySet().stream()
-                .filter(e -> validMessagesTester.test(e.getValue()))
-                .map(e -> new MessageWithId.CreationMessageEntry(e.getKey(), e.getValue()))
-                .map(nuMsg -> createMessageInOutboxAndSend(nuMsg, mailboxSession, outbox, buildMessageIdFunc(mailboxSession, outbox)))
-                .map(msg -> SetMessagesResponse.builder().created(ImmutableMap.of(msg.getCreationId(), msg.getMessage())).build())
-                .reduce(responseBuilder, SetMessagesResponse.Builder::accumulator, SetMessagesResponse.Builder::combiner)
-                .build();
+        List<CreationMessageEntry> creationRequestEntries = request.getCreate();
+        
+        Builder responseBuilder = SetMessagesResponse.builder();
+        creationRequestEntries.stream()
+            .map(entry -> processEntry(entry, mailboxSession))
+            .forEach(either -> either.fold(
+                    error -> responseBuilder.notCreated(error.getCreationId(), error.getValue()),
+                    result -> responseBuilder.created(result.getCreationId(), result.getValue())));
+        return responseBuilder.build();
     }
 
-    private Map<CreationMessageId, SetError> handleCreationErrors(Predicate<CreationMessage> invalidMessagesTester,
-                                                                  SetMessagesRequest request) {
-        return request.getCreate().entrySet().stream()
-                .filter(e -> invalidMessagesTester.test(e.getValue()))
-                .map(e -> new AbstractMap.SimpleEntry<>(e.getKey(), buildSetErrorFromValidationResult(e.getValue().validate())))
-                .collect(Collectors.toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
+    private Either<ErrorWithId, MessageWithId> processEntry(CreationMessageEntry entry, MailboxSession mailboxSession) {
```

Comment:
Way more beautiful than what I proposed in James-1734. Can we get this merged so that I rebase on it and add my logic ?